File: src/lib/events.ts
/**
 * Single source of truth for `app://` event names and their payload shapes.
 *
 * Before this module existed, every emit/listen site used an inline string
 * literal (`"app://settings-updated"`), which made event names impossible to
 * refactor safely and gave no compile-time check on payloads. Keep all event
 * identifiers here; import `EVT` for the name and the matching `*Payload`
 * type when emitting/listening.
 *
 * The Rust side mirrors these names in `src-tauri/src/events.rs`.
 */

export const EVT = {
  // Settings
  SETTINGS_UPDATED: "app://settings-updated",
  // Hotkey / tray action fan-out
  HOTKEY_TRIGGERED: "app://hotkey-triggered",
  // Camera
  CAMERA_ERROR: "app://camera-error",
  CAMERA_REACQUIRE: "app://camera-reacquire",
  // Keyboard display (macOS event tap)
  KEY_PRESSED: "app://key-pressed",
  KEY_RELEASED: "app://key-released",
  ACCESSIBILITY_STATUS: "app://accessibility-status",
  EVENT_TAP_STATUS: "app://event-tap-status",
  // Screen recording — mouse
  MOUSE_DOWN: "app://mouse-down",
  MOUSE_UP: "app://mouse-up",
  MOUSE_MOVE: "app://mouse-move",
  // Screen recording — orchestration
  RECORDING_STATUS: "app://recording-status",
  /** Control-bar → pipeline commands. The pipeline lives in the MAIN camera
   *  window (so camera + screen capture share one document and WebKit's
   *  one-capture-per-page mute policy never fires); the recording bar is a
   *  remote control that sends actions here. */
  RECORDING_CMD: "app://recording-cmd",
  /** Pipeline → control-bar UI state broadcast (status/elapsed/countdown/
   *  info). Emitted on every change and in response to a `sync` command. */
  RECORDING_UI: "app://recording-ui",
  REGION_STARTED: "app://region-started",
  REGION_SELECTED: "app://region-selected",
  REGION_CANCELED: "app://region-canceled",
} as const;

export type EventName = (typeof EVT)[keyof typeof EVT];

// --- Payload shapes (kept loose where payloads are simple) ---

export interface HotkeyPayload {
  action: string;
}

export interface CameraErrorPayload {
  message: string;
}

export interface KeyPayload {
  key: string;
  modifiers: string[];
  timestamp: number;
}

export interface MousePayload {
  /** Physical screen pixels (global). */
  x: number;
  /** Physical screen pixels (global). */
  y: number;
  /** "left" | "right" | "" (move). */
  button: string;
  timestamp: number;
}

export interface RecordingStatusPayload {
  active: boolean;
}

export interface RecordingCmdPayload {
  /** `sync` asks the pipeline to re-broadcast its current UI state (used by
   *  the control bar on mount, since it may open after state was set). */
  action: "toggle" | "toggle-pause" | "clear-info" | "sync";
}

export interface RecordingUiPayload {
  status: "idle" | "countdown" | "recording" | "paused" | "saving";
  elapsed: number;
  countdown: number;
  info: string;
}

export interface PermissionStatusPayload {
  granted: boolean;
}

export interface EventTapStatusPayload {
  active: boolean;
}

File: src-tauri/src/events.rs
//! Single source of truth for `app://` event names on the Rust side.
//!
//! Mirrors `src/lib/events.ts`. Importing these constants instead of inline
//! `&str` literals lets the compiler catch a renamed event at every emit
//! site, and keeps Rust and TypeScript names from drifting.

#[allow(dead_code)]
pub mod evt {
    // Settings
    pub const SETTINGS_UPDATED: &str = "app://settings-updated";
    // Hotkey / tray action fan-out
    pub const HOTKEY_TRIGGERED: &str = "app://hotkey-triggered";
    // Camera
    pub const CAMERA_ERROR: &str = "app://camera-error";
    pub const CAMERA_REACQUIRE: &str = "app://camera-reacquire";
    // Keyboard display (macOS event tap)
    pub const KEY_PRESSED: &str = "app://key-pressed";
    pub const KEY_RELEASED: &str = "app://key-released";
    pub const ACCESSIBILITY_STATUS: &str = "app://accessibility-status";
    pub const EVENT_TAP_STATUS: &str = "app://event-tap-status";
    // Screen recording — mouse
    pub const MOUSE_DOWN: &str = "app://mouse-down";
    pub const MOUSE_UP: &str = "app://mouse-up";
    pub const MOUSE_MOVE: &str = "app://mouse-move";
    // Screen recording — orchestration
    pub const RECORDING_STATUS: &str = "app://recording-status";
    pub const REGION_STARTED: &str = "app://region-started";
    pub const REGION_SELECTED: &str = "app://region-selected";
    pub const REGION_CANCELED: &str = "app://region-canceled";
}
